glibc 2.36 doesn't have libdl.so, it has been melted into libc
Change to libz.so
Index: boost-1.81.0/libs/dll/test/shared_library_load_test.cpp
===================================================================
--- boost-1.81.0.orig/libs/dll/test/shared_library_load_test.cpp
+++ boost-1.81.0/libs/dll/test/shared_library_load_test.cpp
@@ -231,7 +231,7 @@ int main(int argc, char* argv[])
             boost::dll::shared_library("winmm.dll");
             BOOST_TEST(false);
 #elif BOOST_OS_LINUX
-            boost::dll::shared_library("libdl.so");
+            boost::dll::shared_library("libz.so");
             BOOST_TEST(false);
 #endif
         } catch (...) {}
@@ -242,7 +242,7 @@ int main(int argc, char* argv[])
 #if BOOST_OS_WINDOWS
             boost::dll::shared_library("winmm", load_mode::search_system_folders | load_mode::append_decorations);
 #elif BOOST_OS_LINUX
-            boost::dll::shared_library("dl", boost::dll::load_mode::search_system_folders | load_mode::append_decorations);
+            boost::dll::shared_library("z", boost::dll::load_mode::search_system_folders | load_mode::append_decorations);
 #endif
         } catch (...) {
             BOOST_TEST(false);
